File: cpp/test/Ice/interceptor/AMDInterceptorI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <IceUtil/DisableWarnings.h>
#include <AMDInterceptorI.h>
#include <MyObjectI.h>
#include <TestHelper.h>

using namespace std;
using namespace Test;
using namespace Ice;

AMDInterceptorI::AMDInterceptorI(const ObjectPtr& servant) :
    InterceptorI(servant)
{
}

bool
AMDInterceptorI::dispatch(Request& request)
{
    Current& current = const_cast<Current&>(request.getCurrent());

    Context::const_iterator p = current.ctx.find("raiseBeforeDispatch");
    if(p != current.ctx.end())
    {
        if(p->second == "user")
        {
            throw InvalidInputException();
        }
        else if(p->second == "notExist")
        {
            throw ObjectNotExistException(__FILE__, __LINE__);
        }
        else if(p->second == "system")
        {
            throw MySystemException(__FILE__, __LINE__);
        }
    }

    _lastOperation = current.operation;

    if(_lastOperation == "amdAddWithRetry")
    {
        for(int i = 0; i < 10; ++i)
        {
            _lastStatus = _servant->ice_dispatch(request, nullptr, [](exception_ptr ex) {
                try
                {
                    rethrow_exception(ex);
                }
                catch(const MyRetryException&)
                {
                }
                catch(...)
                {
                    test(false);
                }
                return false;
            });
            test(!_lastStatus);
        }

        current.ctx["retry"] = "no";
    }
    else if(current.ctx.find("retry") != current.ctx.end() && current.ctx["retry"] == "yes")
    {
        //
        // Retry the dispatch to ensure that abandoning the result of the dispatch
        // works fine and is thread-safe
        //
        _servant->ice_dispatch(request);
        _servant->ice_dispatch(request);
    }

    _lastStatus = _servant->ice_dispatch(request, []() { return true; }, [this](exception_ptr ex) {
        setException(ex);
        return true;
    });

    p = current.ctx.find("raiseAfterDispatch");
    if(p != current.ctx.end())
    {
        if(p->second == "user")
        {
            throw InvalidInputException();
        }
        else if(p->second == "notExist")
        {
            throw ObjectNotExistException(__FILE__, __LINE__);
        }
        else if(p->second == "system")
        {
            throw MySystemException(__FILE__, __LINE__);
        }
    }

    return _lastStatus;
}

void
AMDInterceptorI::setException(exception_ptr e)
{
    lock_guard lock(_mutex);
    _exception = e;
}

exception_ptr
AMDInterceptorI::getException() const
{
    lock_guard lock(_mutex);
    return _exception;
}

void
AMDInterceptorI::clear()
{
    InterceptorI::clear();
    lock_guard lock(_mutex);
    _exception = nullptr;
}

File: csharp/test/Ice/interceptor/AMDInterceptorI.cs
// **********************************************************************
//
// Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

//
// A dispatch interceptor with special handling for AMD requests
//

using System;

class AMDInterceptorI : InterceptorI, Ice.DispatchInterceptorAsyncCallback
{
    class DispatchInterceptorAsyncCallbackI : Ice.DispatchInterceptorAsyncCallback
    {
        public bool response(bool ok)
        {
            test(ok);
            return false;
        }
        
        public bool exception(System.Exception ex)
        {
            test(ex is Test.RetryException);
            return false;
        }
    };


    internal AMDInterceptorI(Ice.Object servant) : base(servant)
    {
    }
    
    public override Ice.DispatchStatus
    dispatch(Ice.Request request)
    {
        Ice.Current current = request.getCurrent();
        lastOperation_ = current.operation;

        if(lastOperation_.Equals("amdAddWithRetry"))
        {
            for(int i = 0; i < 10; ++i)
            {
                Ice.DispatchInterceptorAsyncCallback cb = new DispatchInterceptorAsyncCallbackI();
                   
                lastStatus_ = servant_.ice_dispatch(request, cb);
                test(lastStatus_ == Ice.DispatchStatus.DispatchAsync);
            }
            
            request.getCurrent().ctx["retry"] = "no";
        }
        

        lastStatus_ = servant_.ice_dispatch(request, this);
        return lastStatus_;
    }

    public bool response(bool ok)
    {
        setActualStatus(ok ? Ice.DispatchStatus.DispatchOK : Ice.DispatchStatus.DispatchUserException);
        return true;
    }

    public bool exception(System.Exception ex)
    {
        setActualStatus(ex);
        return true;
    }

    internal override void
    clear()
    {
        base.clear();
        lock(this)
        {
            actualStatus_ = Ice.DispatchStatus.DispatchAsync;
            exception_ = null;
        }
    }

    internal void 
    setActualStatus(Ice.DispatchStatus status)
    {
        lock(this)
        {
            actualStatus_ = status;
            exception_ = null;
        }
    }

    internal void 
    setActualStatus(System.Exception ex)
    {
        lock(this)
        {
            exception_ = ex;
            actualStatus_ = Ice.DispatchStatus.DispatchAsync;
        }
    }

    internal Ice.DispatchStatus 
    getActualStatus()
    {
        lock(this)
        {
            return actualStatus_;
        }
    }

    internal System.Exception
    getException()
    {
        lock(this)
        {
            return exception_;
        }
    }

    private Ice.DispatchStatus actualStatus_;
    private System.Exception exception_;
}
